app/uk/gov/hmrc/gform/sharedmodel/SmartString.scala

```diff
@@ -0,0 +1,118 @@
+            case Right(expr) => recurse(s"$processed$pre{${interpolations.size}}", post, expr :: interpolations)
+          }
+        case other => JsSuccess((s"$processed$other", interpolations.reverse))
+      }
+
+    recurse("", s, initialInterpolations.reverse)
+  }
+
+  private def readFromTemplateString(value: String) =
+    extractInterpolations(value, Nil)
+      .map { case (s, i) => SmartString(LocalisedString(Map(LangADT.En -> s)), i) }
+
+  private def readLocalisedStringFromTemplateObject(value: Map[String, JsValue]) =
+    LocalisedString.format.reads(JsObject(value.filterKeys(_ =!= interpolationsFieldName)))
+
+  private def readInterpolationsFromTemplateObject(value: Map[String, JsValue]) =
+    value
+      .get(interpolationsFieldName)
+      .map { a =>
+        Reads.of[List[Expr]].reads(a)
+      }
+      .getOrElse(JsSuccess(Nil))
+
+  private def buildSmartStringFromLocalisedStringWithPossibleExpressions(ls: LocalisedString): JsResult[SmartString] =
+    ls.m.foldLeft(JsSuccess(SmartString.empty).asInstanceOf[JsResult[SmartString]]) {
+      case (acc: JsResult[SmartString], (lang: LangADT, s: String)) =>
+        acc.flatMap { ss =>
+          extractInterpolations(s, ss.interpolations).map {
+            case (stringWithPlaceholders, interpolations) =>
+              SmartString(LocalisedString(ss.localised.m + (lang -> stringWithPlaceholders)), interpolations)
+          }
+        }
+    }
+
+  private def buildSmartStringFromLocalisedStringAndInterpolations(
+    ls: LocalisedString,
+    interpolations: List[Expr]): JsResult[SmartString] =
+    buildSmartStringFromLocalisedStringWithPossibleExpressions(ls)
+      .flatMap { ss =>
+        if (interpolations.isEmpty)
+          JsSuccess(ss)
+        else if (ss.interpolations.isEmpty)
+          JsSuccess(ss.copy(interpolations = interpolations))
+        else
+          JsError(s"""|SmartStrings can either be constructed with:
+                      |  embedded expressions (e.g. $${formComponentId}), OR
+                      |  interpolation indexes (e.g. {0}) and a list of interpolation expressions
+                      |but not with both.""".stripMargin)
+      }
+
+  private def readFromTemplateObject(value: Map[String, JsValue]) =
+    for {
+      localisedString <- readLocalisedStringFromTemplateObject(value)
+      interpolations  <- readInterpolationsFromTemplateObject(value)
+      ss              <- buildSmartStringFromLocalisedStringAndInterpolations(localisedString, interpolations)
+    } yield ss
+
+  val templateReads: Reads[SmartString] = new Reads[SmartString] {
+    override def reads(json: JsValue): JsResult[SmartString] = json match {
+      case JsString(value) => readFromTemplateString(value)
+      case obj: JsObject   => readFromTemplateObject(obj.fields.toMap)
+      case _               => JsError(s"Expected a String or an Object while reading a SmartString. Got a ${json.getClass}")
```

Comment:
Maybe output `json` instead of `json.getClass` may help to spot the problem quicker for the user who will see that message